File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int hash[256];
        fill(hash, hash + 256, -1);
        int l=0,r=0;
        int n = s.size();
        int ml = 0;
        while(r < n){
            if(hash[s[r]] != -1){
                if(hash[s[r]] >= l){
                    l = hash[s[r]]+1;
                }
            }

            int len = r-l+1;
            ml = max(ml,len);
            hash[s[r]] = r;
            r++;


        }
        return ml;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        l = 0
        ml = 0

        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[l])
                l+=1
            seen.add(s[right])
            ml = max(ml, right-l+1)

        return ml